### PR TITLE
qcat: use direct memcpy when all the inputs and output share the same scale and zero_point

### PR DESCRIPTION
**Pull request by mingfeima**
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #74999
* #72575
* __->__ #71903

Differential Revision: [D34113897](https://our.internmc.facebook.com/intern/diff/D34113897)

### CODE REVIEW

File: aten/src/ATen/native/quantized/cpu/kernels/QuantizedOpKernels.cpp [hunk 1]
```diff
@@ -41,6 +41,18 @@ void check_tensor_memory_format(const Tensor& ref, const Tensor& other) {
 //
 // Please read the README.md in this directory before editing this file
 
+bool all_inputs_sharing_scale_and_zero_point(
+    const c10::List<Tensor>& qxs,
+    double scale,
+    int64_t zero_point) {
+  bool is_valid = true;
+  for (const auto i : c10::irange(qxs.size())) {
+    is_valid |= qxs[i].q_scale() == scale;
```
Comment (vadimkantorov) [inline aten/src/ATen/native/quantized/cpu/kernels/QuantizedOpKernels.cpp:50]: if `is_valid` is initialized as true, then any further `|=` will not change its value. Did you mean `&=`?

Comment (mingfeima) [reply aten/src/ATen/native/quantized/cpu/kernels/QuantizedOpKernels.cpp]: Ohhh, yes! Fixed.

**Comment by mingfeima** [comment]
TODO: this PR won't cover the scenario from shufflenet since:
* memory format is not channels last since use of `.contiguous()` from channel_shuffle func

**Comment by VitalyFedyunin** [comment]
@VitalyFedyunin has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D34113897).

**Comment by mingfeima** [comment]
### Updates of this PR
originally, qcat would go dequant-cat-quant. This PR checks whether the input has the same qparam with output, if yes use a memcpy instead of dequant/quant.

for performance update, please check the previous PR.

**Comment by jerryzh168** [comment]
is this connected with quantization flow?

**Review by jerryzh168 — APPROVED** [review]

**Comment by mingfeima** [comment]
> is this connected with quantization flow?

it's not related to quantization flow, only a fast path on the quantized kernel.

**Comment by VitalyFedyunin** [comment]
@frank-wei  can you please commandeer this PR internally and take care of landing.

**Comment by frank-wei** [comment]
@frank-wei has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D34113897).

**Pull request closed without merging**